transformer_engine/jax/cpp_extensions/gemm.py

```diff
@@ -0,0 +1,691 @@
+            operand_dtype = jax_dtype_to_te_dtype(lhs_aval.dtype)
+            bias_dtype = jax_dtype_to_te_dtype(bias_aval.dtype)
+            opaque = tex.pack_gemm_descriptor(
+                m,
+                n,
+                k,
+                workspace_size,
+                operand_dtype,
+                jax_dtype_to_te_dtype(out_dtype),
+                bias_dtype,
+                lhs_trans,
+                rhs_trans,
+                fuse_gelu,
+                fuse_bias,
+                grad,
+                accumulate,
+                use_split_accumulator,
+            )
+
+            return custom_caller(
+                CollectiveGemmPrimitive.name,
+                args,
+                opaque,
+                has_side_effect=False,
+                operand_output_aliases=operand_output_aliases,
+            )
+
+    @staticmethod
+    def impl(
+        lhs,
+        lhs_scale_inv,
+        rhs,
+        rhs_scale_inv,
+        bias,
+        gelu_input,
+        out_amax,
+        out_scale,
+        out_dtype,
+        contracting_dims,
+        fuse_gelu,
+        fuse_bias,
+        grad,
+        accumulate,
+        use_split_accumulator,
+    ):
+        assert CollectiveGemmPrimitive.inner_primitive is not None
+
+        (
+            out,
+            out_amax_updated,
+            out_scale_updated,
+            pre_gelu_out,
+            bias_grad,
+            _,
+        ) = CollectiveGemmPrimitive.inner_primitive.bind(
+            lhs,
+            lhs_scale_inv,
+            rhs,
+            rhs_scale_inv,
+            bias,
+            gelu_input,
+            out_amax,
+            out_scale,
+            out_dtype=out_dtype,
+            contracting_dims=contracting_dims,
+            fuse_gelu=fuse_gelu,
+            fuse_bias=fuse_bias,
+            grad=grad,
+            accumulate=accumulate,
+            use_split_accumulator=use_split_accumulator,
+        )
+        return out, out_amax_updated, out_scale_updated, pre_gelu_out, bias_grad
+
+    @staticmethod
+    def batcher(
+        batched_args,
+        batch_dims,
+        *,
+        out_dtype,
+        contracting_dims,
+        fuse_gelu,
+        fuse_bias,
+        grad,
+        accumulate,
+        use_split_accumulator,
+    ):
+        assert CollectiveGemmPrimitive.outer_primitive is not None
+        check_valid_batch_dims(batch_dims)
+        lhs_bdims, *_, bias_bdims, gelu_input_bdims, out_amax_bdims, out_scale_bdims = batch_dims
+
+        # FP8 GEMM only supports non-transposed LHS and transposed RHS
+        lhs, _, rhs, *_ = batched_args
+        lhs_trans = contracting_dims[0] != lhs.ndim - 1
+        rhs_trans = contracting_dims[1] == rhs.ndim - 1
+        lhs = jnp.matrix_transpose(lhs) if lhs_trans and jax_dtype_is_fp8(lhs.dtype) else lhs
+        rhs = jnp.matrix_transpose(rhs) if not rhs_trans and jax_dtype_is_fp8(rhs.dtype) else rhs
+        contracting_dims = (1, 1)
```

Comment:
@denera is there a need to hard-code this?

Reply:
My understanding is that non-transposed LHS and transposed RHS is a cuBlasLt requirement for FP8 GEMM, so this is ensuring that the custom op can do the GEMM even if it receives incorrect layouts. I was planning to add a warning here as well to make it clear that there will be a perf hit.